zetaclient/zeta_supply_checker.go

```diff
@@ -0,0 +1,230 @@
+
+	copyZetaTokenSupplyOnNode := zetaTokenSupplyOnNode
+	copyGenesisAmounts := genesisAmounts
+	nodeAmounts := copyZetaTokenSupplyOnNode.Sub(copyGenesisAmounts)
+	logger.Info().Msgf("--------------------------------------------------------------------------------")
+	logger.Info().Msgf("aborted tx amounts : %s", abortedTxAmounts.String())
+	logger.Info().Msgf("zeta in transit : %s", zetaInTransit.String())
+	logger.Info().Msgf("external chain total supply : %s", externalChainTotalSupply.String())
+	logger.Info().Msgf("effective zeta supply on node : %s", nodeAmounts.String())
+	logger.Info().Msgf("zeta token supply on node : %s", zetaTokenSupplyOnNode.String())
+	logger.Info().Msgf("genesis amounts : %s", genesisAmounts.String())
+	logger.Info().Msgf("eth locked amount : %s", ethLockedAmount.String())
+	if !lhs.Equal(rhs) {
+		logger.Error().Msgf("zeta supply mismatch, lhs : %s , rhs : %s", lhs.String(), rhs.String())
+		return false
+	}
+	logger.Info().Msgf("zeta supply check passed, lhs : %s , rhs : %s", lhs.String(), rhs.String())
+	logger.Info().Msgf("--------------------------------------------------------------------------------")
+	return true
+}
+
+func (zs *ZetaSupplyChecker) AbortedTxAmount() sdkmath.Int {
+	cctxList, err := zs.zetaClient.GetCctxByStatus(types.CctxStatus_Aborted)
+	if err != nil {
+		panic(err)
+	}
+	amount := sdkmath.ZeroUint()
+	for _, cctx := range cctxList {
+		amount = amount.Add(cctx.GetCurrentOutTxParam().Amount)
+	}
+	amountInt, ok := sdkmath.NewIntFromString(amount.String())
+	if !ok {
+		panic("error parsing amount")
+	}
+	return amountInt
+}
+
+func (zs *ZetaSupplyChecker) GetAmountOfZetaInTransit() sdkmath.Int {
+	chainsToCheck := make([]common.Chain, len(zs.externalEvmChain)+1)
+	chainsToCheck = append(append(chainsToCheck, zs.externalEvmChain...), zs.ethereumChain)
+	cctxs := zs.GetPendingCCTXInTransit(chainsToCheck)
+	amount := sdkmath.ZeroUint()
+	for _, cctx := range cctxs {
+		amount = amount.Add(cctx.GetCurrentOutTxParam().Amount)
+	}
+	amountInt, ok := sdkmath.NewIntFromString(amount.String())
+	if !ok {
+		panic("error parsing amount")
+	}
+	return amountInt
+}
+func (zs *ZetaSupplyChecker) GetPendingCCTXInTransit(receivingChains []common.Chain) []*types.CrossChainTx {
+	cctxInTransit := make([]*types.CrossChainTx, 0)
+	for _, chain := range receivingChains {
+		cctx, err := zs.zetaClient.GetAllPendingCctx(chain.ChainId)
+		if err != nil {
+			continue
+		}
+		nonceToCctxMap := make(map[uint64]*types.CrossChainTx)
+		for _, c := range cctx {
+			if c.GetInboundTxParams().CoinType == common.CoinType_Zeta {
+				nonceToCctxMap[c.GetCurrentOutTxParam().OutboundTxTssNonce] = c
+			}
+		}
+
+		trackers, err := zs.zetaClient.GetAllOutTxTrackerByChain(chain, Ascending)
+		if err != nil {
+			continue
+		}
+		for _, tracker := range trackers {
+			zs.logger.Info().Msgf("tracker exists for nonce: %d , removing from supply checks", tracker.Nonce)
+			delete(nonceToCctxMap, tracker.Nonce)
```

Comment:
the purpose here is to exclude pending cctcs who have a tracker?

Reply:
Yes